### ISSUE TITLE
Android Pie - Bluetooth controls and metadata

### ISSUE THREAD

**Issue by glennguy** [daneren2005/Subsonic#878]
Hi @daneren2005 

I'm running dsub on a pixel and after upgrading Android to 9.0 there are temperamental issues with displaying metadata and also remote control. I'm noticing that sometimes there is no metadata displayed at all and this is when the remote controls stop working as well. Others are having problems with other media players but any solutions I've come across haven't worked.

However, most of the time the remote controls work and I do get *some* metadata. What I've seen is that when starting a new album/playlist the metadata appears instantly with the correct information, but once the next track starts or if you skip a track, the actual track name populates in both the track name and artist name fields, and the album name is empty. It also takes 3-4 seconds for the information to be updated. So it works but it doesn't work -

I've tested Google Play Music and Spotify but both are working correctly, so I'm hoping there's a solution to this. I've had a peek at Spotify in Jadx but the obfuscated names are a bit too hard to sift through.

I have tried a few things with the code to narrow down the problem:

* Compiling against API 28

This seemed to be fairly easy. Apart from a few changes to the gradle files (adding flavorDimensions and settings buildTypes in ServerProxy) and proguard.cfg (adding a dontwarn line) the only thing that stopped the app from working is lack of HTTPS which is worked around by adding `android:usesCleartextTraffic="true">` in the Application tag in the manifest. I can make a PR if you want. No improvement though.

* Upgrading the deprecated MediaSession etc. in Util and RemoteControlClientLP to their 'Compat' versions.

This didn't seem to do anything either.

* Playing with the broadcastNewTrackInfo and broadcastPlaybackStatusChange methods in Util

Commenting out `context.sendBroadcast(avrcpIntent)` doesn't seem to affect anything... I found these [changes](https://github.com/aosp-mirror/platform_packages_apps_music/commit/f02d3c605f5d1d091ca7e27170010ad23cd22134#diff-29edfb2901ba2915d7e6f9c2d9e4ad9d) in com.android.music that might have now made their way into Pie, if so that could explain the change in behavior.

There's some parts of the program that I still haven't figured out the workings of, but I'm keen to help so let me know if there's any other info you want or if you have any suggestions I could try.

Thanks.

**Comment by QuillOmega0** [comment]
+1 to this I found the same issues on Android Pie not sending MetaData through.

**Comment by wljohnson05** [comment]
Same issue here with me the other day in my truck on Pie. I didn't dig into it, but for a while it was blank, and then it starting showing up.

**Comment by QuillOmega0** [comment]
Also Glennguy it appears that the magic is happening in https://github.com/daneren2005/Subsonic/blob/edge/app/src/main/java/github/daneren2005/dsub/util/compat/RemoteControlClientLP.java for sending MetaData updates

**Comment by glennguy** [comment]
Ok so I've returned for another look at whats going on and a bunch of lines like this in my log: `D/NewAvrcpMediaPlayerWrapper: github.daneren2005.dsub: getCurrentMetadata: Using playlist data: { mediaId="NowPlayingId1605450" title="When I'm a Hundred Sixty-Four" artist="When I'm a Hundred Sixty-Four" album="" duration=0 trackPosition=2/20 }`. 

I've found in the android source that (apart from seeming that the entire avrcp module has been rewritten) it is using playlist/queue data for metadata rather than getting it from the actual player. That change was added on 10th of May 2018, and removed again 5 days ago - 10th September -- see [here](https://android.googlesource.com/platform/packages/apps/Bluetooth/+/725986178cbbfb11dd99cad1ebb4ea6f7aa1b1b2%5E%21/src/com/android/bluetooth/newavrcp/MediaPlayerWrapper.java) and [here](https://android.googlesource.com/platform/packages/apps/Bluetooth/+/48db51dd1fc3cbd45d322b9b7a93ee6d4c3d1d63%5E%21/) Clearly it made it in for Android 9.0, hopefully this revert will fix this part problem and be a part of the next monthly update, we will see.

@QuillOmega0 @wljohnson05 are you seeing song name for both title and artist?

**Comment by jeffgt14** [comment]
I was thinking it was just Android 9 that was giving me fits but I'm having the exact same problem described here. Other players seem to work fine.

**Comment by QuillOmega0** [comment]
Yep. Most of the time I get the Title of the song in the title and artist
field. Other players are fine in that regard. This happened since the Pie
beta.

**Comment by glennguy** [comment]
I've found a solution for the issue with the metadata, still intermittent with the remote control side of things though. Opening another app (or switching music app in android Auto) and switching back seems to 'reset' it and the connection is made. Still investigating.
